### PR TITLE
[API] Refactor converting arguments of commands from IRIs to identifiers

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | https://github.com/Sylius/Sylius/pull/12487
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/spec/Serializer/CommandFieldItemIriToIdentifierDenormalizerSpec.php [hunk 1]
```diff
@@ -41,91 +39,78 @@ function let(
...
-                'email' => 'john@example.com',
-            ],
-                AddProductReview::class,
-                null,
-                $context
-            )
-            ->shouldReturn($addProductReview)
-        ;
-    }
+//    function it_supports_denormalization_add_product_review(): void {
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/spec/Serializer/CommandFieldItemIriToIdentifierDenormalizerSpec.php:45]: Commented for tests

File: src/Sylius/Bundle/ApiBundle/Exception/NoRouteMatches.php [hunk 2]
```diff
@@ -11,12 +11,10 @@
...
+namespace Sylius\Bundle\ApiBundle\Exception;
 
-interface CommandItemIriArgumentToIdentifierMapInterface
-{
-    /** @psalm-var class-string $className */
-    public function get(string $className): string;
+use ApiPlatform\Core\Exception\InvalidArgumentException;
 
-    public function has(?string $className): bool;
+final class NoRouteMatches extends InvalidArgumentException
```
Comment (stloyd) [inline src/Sylius/Bundle/ApiBundle/Exception/NoRouteMatches.php:18]: Still this should have "exception" in name IMHO.

File: src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php [hunk 3]
```diff
@@ -30,35 +31,64 @@ final class CommandFieldItemIriToIdentifierDenormalizer implements ContextAwareD
...
     }
 
     public function supportsDenormalization($data, $type, $format = null, array $context = [])
     {
-        return $this->commandItemIriArgumentToIdentifierMap->has($this->getInputClassName($context));
+        /** @psalm-var class-string $inputClassName */
+        $inputClassName = $this->getInputClassName($context);
+
+        /** @var array $classInterfaces */
+        $classInterfaces = class_implements($inputClassName);
```
Comment (stloyd) [inline src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php:50]: Why assign & not directly in loop setup?

File: src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php [hunk 4]
```diff
@@ -30,35 +31,64 @@ final class CommandFieldItemIriToIdentifierDenormalizer implements ContextAwareD
...
-        $this->commandItemIriArgumentToIdentifierMap = $commandItemIriArgumentToIdentifierMap;
     }
 
     public function supportsDenormalization($data, $type, $format = null, array $context = [])
     {
-        return $this->commandItemIriArgumentToIdentifierMap->has($this->getInputClassName($context));
+        /** @psalm-var class-string $inputClassName */
+        $inputClassName = $this->getInputClassName($context);
+
+        /** @var array $classInterfaces */
```
Comment (stloyd) [inline src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php:49]:
```suggestion
        /** @var array<int, string> $classInterfaces */
```

File: src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php [hunk 5]
```diff
@@ -30,35 +31,64 @@ final class CommandFieldItemIriToIdentifierDenormalizer implements ContextAwareD
...
     {
         /** @psalm-var class-string $inputClassName */
         $inputClassName = $this->getInputClassName($context);
 
-        $fieldName = $this->commandItemIriArgumentToIdentifierMap->get($inputClassName);
-
-        if (array_key_exists($fieldName, $data)) {
-            $data[$fieldName] = $this->itemIriToIdentifierConverter->getIdentifier($data[$fieldName]);
+        /** @var array $classInterfaces */
+        $classInterfaces = class_implements($inputClassName);
```
Comment (stloyd) [inline src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php:67]: Same about assign & phpdoc

File: src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php [hunk 6]
```diff
@@ -30,35 +31,64 @@ final class CommandFieldItemIriToIdentifierDenormalizer implements ContextAwareD
...
+            if ($classInterface === CommandFieldItemIriToIdentifierAwareInterface::class) {
+                $classFields = get_class_vars($inputClassName);
+                foreach ($classFields as $classFieldName => $classFieldValue) {
+                    $serializedFieldName = null;
+
+                    if ($this->isConvertableField($classFieldName)) {
+                        $serializedFieldName = substr($classFieldName, 0, -4);
+                    }
+
+                    if (array_key_exists($serializedFieldName, $data)) {
```
Comment (stloyd) [inline src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php:79]:
```suggestion
                    if (isset ($data[$serializedFieldName])) {
```
Or even "return early".

File: src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php [hunk 7]
```diff
@@ -30,35 +31,61 @@ final class CommandFieldItemIriToIdentifierDenormalizer implements ContextAwareD
...
     {
         /** @psalm-var class-string $inputClassName */
         $inputClassName = $this->getInputClassName($context);
 
-        $fieldName = $this->commandItemIriArgumentToIdentifierMap->get($inputClassName);
-
-        if (array_key_exists($fieldName, $data)) {
-            $data[$fieldName] = $this->itemIriToIdentifierConverter->getIdentifier($data[$fieldName]);
+        foreach (class_implements($inputClassName) as $classInterface) {
+            if ($classInterface === CommandFieldItemIriToIdentifierAwareInterface::class) {
```
Comment (stloyd) [inline src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php:68]: I would go with negation & calling `continue;` instead so deep if indent.

Just my 2 cents ;)

File: src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php [hunk 8]
```diff
@@ -74,4 +101,8 @@ private function getInputClassName(array $context): ?string
     {
         return $context['input']['class'] ?? null;
     }
+
+    private function isConvertableField(string $classFieldName): bool {
+        return substr($classFieldName, -4) === 'Code';
```
Comment (stloyd) [inline src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php:106]: This check is quite really fragile IMHO;)

File: src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php [hunk 9]
```diff
@@ -30,35 +31,61 @@ final class CommandFieldItemIriToIdentifierDenormalizer implements ContextAwareD
...
         /** @psalm-var class-string $inputClassName */
         $inputClassName = $this->getInputClassName($context);
 
-        $fieldName = $this->commandItemIriArgumentToIdentifierMap->get($inputClassName);
-
-        if (array_key_exists($fieldName, $data)) {
-            $data[$fieldName] = $this->itemIriToIdentifierConverter->getIdentifier($data[$fieldName]);
+        foreach (class_implements($inputClassName) as $classInterface) {
+            if ($classInterface === CommandFieldItemIriToIdentifierAwareInterface::class) {
+                foreach (get_class_vars($inputClassName) as $classFieldName => $classFieldValue) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Serializer/CommandFieldItemIriToIdentifierDenormalizer.php:69]: I would be for iterating on data from the request and checking if there is an iri field, and that field should be converted to identifier

**Review by GSadee — DISMISSED** [review]

**Review by SirDomin — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Command/Checkout/ChooseShippingMethod.php [hunk 10]
```diff
@@ -13,11 +13,12 @@
 
 namespace Sylius\Bundle\ApiBundle\Command\Checkout;
 
+use Sylius\Bundle\ApiBundle\Command\IriToIdentifierConversionAwareInterface;
 use Sylius\Bundle\ApiBundle\Command\OrderTokenValueAwareInterface;
 use Sylius\Bundle\ApiBundle\Command\SubresourceIdAwareInterface;
 
 /** @experimental */
-class ChooseShippingMethod implements OrderTokenValueAwareInterface, SubresourceIdAwareInterface
+class ChooseShippingMethod implements OrderTokenValueAwareInterface, SubresourceIdAwareInterface, IriToIdentifierConversionAwareInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Command/Checkout/ChooseShippingMethod.php:21]: I'm wondering if both SubresourceIdAwareInterface and IriToIdentifierConversionAwareInterface inferfaces shouldn't be extracted to one. Same for most of usage of IriToIdentifierConversionAwareInterface

**Review by lchrusciel — APPROVED** [review]
I'm wondering if we shouldn't

**Review by AdamKasp — APPROVED** [review]

**Comment by AdamKasp** [comment]
Thanks, Tomasz! :1st_place_medal:

**Pull request merged**